src/http_client_async.zig

```diff
@@ -772,6 +864,184 @@ pub const HTTPThread = struct {
+
+const INET6_ADDRSTRLEN = if (bun.Environment.isWindows) 65 else 46;
+
+fn canonicalizeIP(addr_str: []const u8, outIP: *[INET6_ADDRSTRLEN + 1]u8) ![]const u8 {
+    if (addr_str.len >= INET6_ADDRSTRLEN) {
+        return error.Invalid;
+    }
+    var ip_std_text: [INET6_ADDRSTRLEN + 1]u8 = undefined;
+    // we need a null terminated string as input
+    bun.copy(u8, outIP, addr_str);
+    outIP[addr_str.len] = 0;
+
+    var af: c_int = std.os.AF.INET;
+    // get the standard text representation of the IP
+    if (c_ares.ares_inet_pton(af, outIP, &ip_std_text) != 1) {
+        af = std.os.AF.INET6;
+        if (c_ares.ares_inet_pton(af, outIP, &ip_std_text) != 1) {
+            return error.Invalid;
+        }
+    }
+    // ip_addr will contain the null-terminated string of the cannonicalized IP
+    if (c_ares.ares_inet_ntop(af, &ip_std_text, outIP, outIP.len) == null) {
+        return error.Invalid;
+    }
+    // use the null-terminated size to return the string
+    const size = bun.len(bun.cast([*:0]u8, outIP));
+    return outIP[0..size];
+}
+
+fn ipv4ToString(ipv4Bytes: *[4]u8, outIP: *[INET6_ADDRSTRLEN + 1]u8) ![]const u8 {
+    var ipv4String: [16]u8 = undefined;
+    return try canonicalizeIP(try std.fmt.bufPrint(&ipv4String, "{}.{}.{}.{}", .{ ipv4Bytes[0], ipv4Bytes[1], ipv4Bytes[2], ipv4Bytes[3] }), outIP);
+}
+
+const hexChars = "0123456789ABCDEF";
+fn ipv6ToString(ipv6Bytes: *[16]u8, outIP: *[INET6_ADDRSTRLEN + 1]u8) ![]const u8 {
+    var index: usize = 0;
+    bun.copy(u8, outIP, ipv6Bytes);
+    outIP[ipv6Bytes.len] = 0;
+
+    for (ipv6Bytes, 0..) |byte, i| {
+        if (i > 0 and i % 2 == 0) {
+            outIP[index] = ':'; // Add a colon every 2 bytes
+            index += 1;
+        }
+        outIP[index] = hexChars[byte >> 4];
+        index += 1;
+        outIP[index] = hexChars[byte & 0x0F];
+        index += 1;
+    }
+
+    return try canonicalizeIP(outIP[0..index], outIP);
+}
+pub fn onCertError(
+    client: *HTTPClient,
+    comptime is_ssl: bool,
+    socket: NewHTTPContext(is_ssl).HTTPSocket,
+    certError: HTTPCertError,
+) bool {
+    if (comptime is_ssl == false) {
+        @panic("onCertError called on non-ssl socket");
+    }
+    if (client.reject_unauthorized) {
+        const ssl_ptr = @as(*BoringSSL.SSL, @ptrCast(socket.getNativeHandle()));
+        if (BoringSSL.SSL_get_peer_cert_chain(ssl_ptr)) |cert_chain| {
+            if (BoringSSL.sk_X509_value(cert_chain, 0)) |x509| {
+
+                // check if we need to report the error (probably to `checkServerIdentity` was informed from JS side)
+                // this is the slow path
+                if (client.signals.get(.cert_errors)) {
+                    // clone the relevant data
+                    const cert_size = BoringSSL.i2d_X509(x509, null);
+                    var cert = bun.default_allocator.alloc(u8, @intCast(cert_size)) catch @panic("OOM");
+                    var buffer_ptr = @as([*c]u8, cert.ptr);
```

Comment:
we should avoid creating any cases to `[*c]u8` in our code

Reply:
I will take a look and change the translated.zig instead of using this convertions.

Reply:
fixed